common/persistence/history_manager.go

```diff
@@ -674,36 +724,29 @@ func (m *executionManagerImpl) readHistoryBranch(
 			return historyEvents, historyEventBatches, nil, dataSize, serviceerror.NewDataLoss(fmt.Sprintf("corrupted history event batch, empty events"))
 		}
 
-		firstEvent := events[0]           // first
-		eventCount := len(events)         // length
-		lastEvent := events[eventCount-1] // last
-
-		if firstEvent.GetVersion() != lastEvent.GetVersion() || firstEvent.GetEventId()+int64(eventCount-1) != lastEvent.GetEventId() {
-			// in a single batch, version should be the same, and ID should be contiguous
-			m.logger.Error("Corrupted event batch",
-				tag.FirstEventVersion(firstEvent.GetVersion()), tag.WorkflowFirstEventID(firstEvent.GetEventId()),
-				tag.LastEventVersion(lastEvent.GetVersion()), tag.WorkflowNextEventID(lastEvent.GetEventId()),
-				tag.Counter(eventCount))
-			return historyEvents, historyEventBatches, nil, dataSize, serviceerror.NewDataLoss("corrupted history event batch, wrong version and IDs")
+		if !request.ReverseOrder {
+			err = m.validateEventsBatch(events, token.LastEventID)
+			if err != nil {
+				return historyEvents, historyEventBatches, nil, dataSize, err
+			}
 		}
-		if firstEvent.GetEventId() != token.LastEventID+1 {
-			m.logger.Error("Corrupted non-contiguous event batch",
-				tag.WorkflowFirstEventID(firstEvent.GetEventId()),
-				tag.WorkflowNextEventID(lastEvent.GetEventId()),
-				tag.TokenLastEventID(token.LastEventID),
-				tag.Counter(eventCount))
-			return historyEvents, historyEventBatches, nil, dataSize, serviceerror.NewDataLoss("corrupted history event batch, eventID is not contiguous")
+
+		if request.ReverseOrder {
+			// batch of events is still sorted in normal order, have to reverse it
+			sort.Slice(events, func(i, j int) bool {
+				return events[i].EventId > events[j].EventId
+			})
 		}
```

Comment:
this additional cost is not ok in production env, especially when there are lots of call to reverse dump history 
let us try to create a dedicated code path for reverse dump history, plus the logic is different